### ISSUE TITLE
Proposed enhancement to enum_name

### ISSUE THREAD

**Issue by ga2k** [Neargye/magic_enum#40]
_EDIT: Thinking a bit harder, it might be better that this functionality be moved to a new method, perhaps called_ `enum_name_of_bits_set` _or something and leave enum_name alone..._

Hi, I have no clue about github or pull requests, so I thought I'd write my proposed enhancement to enum_name here...

If an enum contains bitfield values, there exists already the ability to bitwise AND and OR the values together to create flag fields. But there is no way to print such a flag.

Example : 
```
using namespace magic_enum;
using namespace magic_enum::bitwise_operators;

enum Rooms { None = 0, Bedroom = 1 << 0, Lounge = 1 << 1, Kitchen = 1 << 2 };
Rooms toPaint = Rooms::Bedroom | Rooms::Kitchen;

std::cout << "My wife want me to paint these : " << enum_name<Rooms>(toPaint) << '\n';
```
Desired output :
**My wife want me to paint these : Bedroom|Kitchen**

Actual output :
**My wife want me to paint these :**

Here is a replacement

````
    // Returns string enum name from enum value.
    // If enum value does not have name or value out of range, returns empty string.
    template <typename E>
    [[nodiscard]] constexpr auto enum_name(E value) noexcept 
    -> detail::enable_if_enum_t<E, std::string_view>
    {
        //  Begin Changes ga2k 2020-05-29

        //  return enum_traits<E>::name(value);

        auto n = enum_traits<E>::name(value);
        if (!n.empty())
            return n;

        //  See if there is a name for each bit set in value. 
        //  If so, concatenate them with '|' separator

        //  First find largest value

        constexpr auto values = enum_values<E>();
        constexpr auto largest = values[values.size() - 1];
        static std::string s;

        for (int b = 0; (1 << b) <= largest; ++b)
        {
            if ((value & (1 << b)) == (1 << b))	// This bit is set in value
            {
                auto n2 = enum_traits<E>::name(static_cast<E>(1 << b));

                if (n2.empty())
                    return n2;	// No name at this bt value : not a bitfield value, so bail

                if (!s.empty())
                {
                    s += '|';
                }

                s += n2;
            }
        }

        return static_cast<std::string_view>(s);
            
        // End Additions ga2k 2020-05-29
    }
````

Thoughts?

**Comment by Neargye** [comment]
Hi @ga2k,

Sorry for a long answer, adding full flags functionality requires more time. 

While we can use this function, it works the way you want.

```cpp
template <typename E>
[[nodiscard]] auto enum_flag_name(E value) noexcept -> std::enable_if_t<std::is_enum_v<std::decay_t<E>>, std::string> {
  using namespace magic_enum;
  using D = std::decay_t<E>;
  using U = std::underlying_type_t<D>;

  if (const auto n = enum_name<D>(value); !n.empty()) {
    return std::string{n};
  }

  auto name = std::string{};
  for (int i = 0; i < std::numeric_limits<U>::digits; ++i) {
    if (const auto v = (0x1 << i); (static_cast<U>(value) & v) != 0) {
      if (const auto n = enum_name<D>(static_cast<D>(v)); !n.empty()) {
        if (name.empty()) {
          name = n;
        } else {
          name.append(1, '|').append(n);
        }
      } else {
        return {}; // Value out of range.
      }
    }
  }

  return name;
}
```

**Comment by Vennor** [comment]
Thank you both for this! I'd really like to see it implemented in the library.

**Comment by Neargye** [comment]
Hi @Vennor, 
I work on full enum flag api, it's need to add enum-flag-value and enum-flag-name enum-flag-cast and etc.

**Comment by ga2k** [comment]
Thanks for that. I'll try it out and see how it goes.

One thing I noticed earlier, when working with flags, is the enormously 
large arrays it creates. For example. an enum with 16 bit values creates 
an array behind it 65535 elements long. Obviously, if every bit 
combination was possible, this should be the size of the array. But in 
actual use the bit combinations are many many times smaller, and so the 
array is extremely sparse. It seems I can only have 9 bits in a bitfield 
before MSVC says the program is too complex to compile.

fatal error C1026: parser stack overflow, program too complex

I tried adding the compiler flags /constexpr:depth65535 and 
/constexpr:steps65535 but nothing changed.

I've considered looking into a mod to the code to use an unordered_map  
instead of an array, but frankly the code scares me and I don't know 
what unintended consequences there might be.

Cheers,
Geoffrey.

**Comment by ga2k** [comment]
Hi, that all works as expected. Thanks a heap.

But, there is always more, isn't there?

enum_cast can't handle converting an int  with a bitfield or a string with a bitfield back into an enum.

Here's an example. I've made the output mirror the code as closely as I could.

```
#include <iostream>
#include <magic_enum.hpp>

using namespace std;
using namespace magic_enum;
using namespace magic_enum::bitwise_operators;

int main()
{
	enum class BF { None = 0, RESET = 0x001, STOP = 0x002, GO = 0x004 };
	cout << "enum class BF { None = 0, RESET = 0x001, STOP = 0x002, GO = 0x004 };\n\n";

	BF bf = BF::RESET | BF::GO;
	cout << "BF bf = BF::RESET | BF::GO" << "\n\n";

	string s = enum_flag_name<BF>(bf);
	cout << "string s = enum_flag_name<BF>(bf)              //     s is " << s << "\n";

	BF bf_s = enum_cast<BF>(s).value_or(BF::None);
	cout << "BF  bf_s = enum_cast<BF>(s).value_or(BF::None) //  bf_s is " << enum_flag_name<BF>(bf_s) << "\n";
	cout << "                                               // expected " << s << "\n\n";

	int i = enum_integer<BF>(bf);
	cout << "int    i = enum_integer<BF>(bf)                //     i is " << i << "\n";

	BF bf_i = enum_cast<BF>(i).value_or(BF::None);
	cout << "BF  bf_i = enum_cast<BF>(i).value_or(BF::None) //  bf_i is " << enum_flag_name<BF>(bf_i) << "\n";
	cout << "                                               // expected " << s << "\n\n";

	return 0;
}
```
and this is the output
```
enum class BF { None = 0, RESET = 0x001, STOP = 0x002, GO = 0x004 };

BF bf = BF::RESET | BF::GO

string s = enum_flag_name<BF>(bf)              //     s is RESET|GO
BF  bf_s = enum_cast<BF>(s).value_or(BF::None) //  bf_s is None
                                               // expected RESET|GO

int    i = enum_integer<BF>(bf)                //     i is 5
BF  bf_i = enum_cast<BF>(i).value_or(BF::None) //  bf_i is None
                                               // expected RESET|GO
```
I think that fixing this will be A LOT more work than enum_flag_name was...

**Comment by Neargye** [comment]
@Vennor  @ga2k, I work on flags, you can play with it in branches dev https://github.com/Neargye/magic_enum/tree/dev

**Comment by Neargye** [comment]
Which name `enum_flag_name` or `enum_flags_name`? `enum_name_flag` or `enum_name_flags`?
Or merge with `enum_name` and not make a separate function ?

**Comment by Vennor** [comment]
I think it depends a lot on the details. As far as I remember magic_enum doesn't support multiple keys for the same value. But where are flags in this? Consider the code below.
```cpp
enum class color
{
  red     = 1 << 0,
  green   = 1 << 1,
  blue    = 1 << 2,
  yellow  = red | green,
  magenta = red | blue,
  cyan    = green | blue
};
```
I imagine that the output of  `enum_name_flag(color::yellow)` will be `red|green`. What if I wanted to get `yellow` though? I need another way to do this, like `enum_name(color::yellow)`. This, I believe, is the main reason for it to be a separate function.

As for the name I'd go with `enum_name_flag`. Using `name` as the second word differentiates it from `enum_type_name` and brings it semantically closer to `enum_name` which for me means it's a variation of the same functionality. Using a singular form `flag` suggests that a single value is returned as opposed to an array of values as is the case with `enum_names`.

**Comment by Neargye** [comment]
@Vennor  Thanks for your review.  

As for the aliases, need to recheck all API, it may even be easier to create a separate namespace with duplicate API but with different behavior for flag. 

For example
```cpp
enum class color
{
  red     = 1 << 0,
  green   = 1 << 1,
  blue    = 1 << 2,
  yellow  = red | green,
  magenta = red | blue,
  cyan    = green | blue
};
```
* `magic_enum::enum_count<color>() -> 6`
* `magic_enum::flags::enum_count<color>() -> 3`  or `magic_enum::enum_flag_count<color>() -> 3`

**Comment by ga2k** [comment]
> Which name `enum_flag_name` or `enum_flags_name`? `enum_name_flag` or `enum_name_flags`?
> Or merge with `enum_name` and not make a separate function ?

I like enum_name_something, I think it flow better with the thinking process of a name.

I like enum_name_flags, because you are likely to have multiple ones set. If you don't, using enum_name does the same thing.

**Comment by ga2k** [comment]
You make a good argument for enum_name_flag. I hadn't considered the other "..s" cases, which return an array. It should definitely not indicate via mnemonics that it'll return an array.

So I change my vote to enum_name_flag.

But I wouldn't hate it to be integrated into a single enum_name function. Semantically, it makes just as much sense to be there. It's returning a name. It just happens to be a flag, or multiple flags. The user will know which is which.

Cheers,
Geoff.

> I think it depends a lot on the details. As far as I remember magic_enum doesn't support multiple keys for the same value. But where are flags in this? Consider the code below.
> 
> ```c++
> enum class color
> {
>   red     = 1 << 0,
>   green   = 1 << 1,
>   blue    = 1 << 2,
>   yellow  = red | green,
>   magenta = red | blue,
>   cyan    = green | blue
> };
> ```
> 
> I imagine that the output of `enum_name_flag(color::yellow)` will be `red|green`. What if I wanted to get `yellow` though? I need another way to do this, like `enum_name(color::yellow)`. This, I believe, is the main reason for it to be a separate function.
> 
> As for the name I'd go with `enum_name_flag`. Using `name` as the second word differentiates it from `enum_type_name` and brings it semantically closer to `enum_name` which for me means it's a variation of the same functionality. Using a singular form `flag` suggests that a single value is returned as opposed to an array of values as is the case with `enum_names`.

**Comment by ga2k** [comment]
As for

`color c(color::yellow)`

Method 1:

`enum_name_flag` could look and see if there is a enum by name, not by value. It would look for an actual enum named `yellow` and return that. Otherwise look for an alias `yellow` and return the component flags `red|green`

This would be many times slower than a value search, though.

Method 2:

Something can be done in the enum's traits like `enum_range` does. Something like

Method 2A
```
enum_name_returns_alias<color>(true);
enum_name_flag(c) returns yellow,

enum_name_returns_alias<color>(false);
enum_name_flag(c) returns red|green.
```
OR

Method 2B

extend `enum_name_flag` (or `enum_name` in a unified version) to accept a `bool` that says whether we want the alias or the flags

----------

I would go for Method 2B if using a single `enum_name` for both types of operation, or Method 1 for `enum_name_flag` if using separate `enum_name` and `enum_name_flag` functions.

Cheers,
Geoff

> I think it depends a lot on the details. As far as I remember magic_enum doesn't support multiple keys for the same value. But where are flags in this? Consider the code below.
> 
> ```c++
> enum class color
> {
>   red     = 1 << 0,
>   green   = 1 << 1,
>   blue    = 1 << 2,
>   yellow  = red | green,
>   magenta = red | blue,
>   cyan    = green | blue
> };
> ```
> 
> I imagine that the output of `enum_name_flag(color::yellow)` will be `red|green`. What if I wanted to get `yellow` though? I need another way to do this, like `enum_name(color::yellow)`. This, I believe, is the main reason for it to be a separate function.
> 
> As for the name I'd go with `enum_name_flag`. Using `name` as the second word differentiates it from `enum_type_name` and brings it semantically closer to `enum_name` which for me means it's a variation of the same functionality. Using a singular form `flag` suggests that a single value is returned as opposed to an array of values as is the case with `enum_names`.

**Comment by Neargye** [comment]
Update dev branch, can play with, and report on the problems you find.

Example:

```cpp
enum class color
{
  red     = 1 << 0,
  green   = 1 << 1,
  blue    = 1 << 2,
  yellow  = red | green,
  magenta = red | blue,
  cyan    = green | blue
};
```
```cpp
using namespace magic_enum;
enum_count<color>() -> 6
enum_name(color::yellow) -> yellow
enum_cast(color::red | color::blue | color::green) -> std::nullopt; 
```
```cpp
using namespace magic_enum::flags;
enum_count<color>() -> 3
enum_name(color::yellow) -> red | green
enum_cast(color::red | color::blue | color::green) -> color::red | color::blue | color::green; 
```

**Comment by ga2k** [comment]
Hey, thanks! I'll try it and let you know.

**Comment by Neargye** [comment]
Did you have time to view the new flags-API? @Vennor @ga2k 
I'm wondering if a separate namespace is OK for flags-api?

**Comment by ga2k** [comment]
> Did you have time to view the new flags-API? @Vennor @ga2k
> I'm wondering if a separate namespace is OK for flags-api?

This is working perfectly. And better, I've tried up to 16 flags at once, and that works, too. I don't know if that is because of code changes, or MSVS doing better at compiling, but it's great news.

I tried enum_cast from an integer, and it worked as expected. Casting from a std::string says "Not implemented" so I guess that's in the works.

I prefer it in it's own namespace "flags", keeps things better separated and forces me to think what I'm doing.

Keep up the great work.
